### PR TITLE
Hoover if it was working

### PR DESCRIPTION
**Pull request by kaplun**

### CODE REVIEW

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 1]
```diff
@@ -790,39 +790,43 @@ def get_all_paper_data_of_author(pid):
     return _select_from_aidpersonidpapers_where(select=['personid', 'bibref_table', 'bibref_value', 'bibrec', 'flag'], pid=pid)
 
 
-def get_papers_of_author(pid, claimed_only=False, include_rejected=False):  # get_all_paper_records
+def get_papers_of_author(pid, include_claimed=True, include_unclaimed=True, include_rejected=False):   ### get_all_paper_records
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:793]: @petros-ioannidis you are changing the signature of `claimed_only=False`, `include_claimed=True, include_unclaimed=True`, but the function is used elsewhere (e.g. in `orcidutils`). Please, upgrade everywhere...
cc @MSusik

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 2]
```diff
@@ -1785,7 +1789,7 @@ def remove_arxiv_papers_of_author(pid):
             'where tag=%s and personid=%s', ('arxiv_papers', pid))
 
 
-def _add_external_id_to_author(pid, ext_sys, ext_id):  # add_personID_external_id
+def add_external_id_to_author(pid, ext_sys, ext_id, avoid_id_duplication=True):   ### add_personID_external_id
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:1792]: @petros-ioannidis again change of function name. Have you checked if it is ok elsewhere?

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 3]
```diff
@@ -2280,7 +2290,7 @@ def get_author_data_associations(table_name="`aidPERSONIDDATA`"):  # get_full_pe
                    'from %s' % table_name)
 
 
-def _get_inspire_id_of_author(pid):  # get_inspire_ids_by_pids
+def get_inspire_id_of_author(pid):   ### get_inspire_ids_by_pids
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2293]: @petros-ioannidis ditto

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 4]
```diff
@@ -2290,8 +2300,14 @@ def _get_inspire_id_of_author(pid):  # get_inspire_ids_by_pids
...
     '''
-    return _select_from_aidpersoniddata_where(select=['data'], pid=pid, tag='extid:INSPIREID')
-
+    result =  _select_from_aidpersoniddata_where(select=['data'], pid=pid, tag='extid:INSPIREID')
+    ##WRONG exception here
+    if result:
+        if len(result) > 1:
+            from invenio.bibauthorid_hoover_exceptions import MultipleIdsOnSingleAuthorException
+            raise MultipleIdsOnSingleAuthorException('Conflict in aidPERSONIDDATA', pid, 'INSPIREID', result)
+        return result[0][0]
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2309]: @petros-ioannidis here you are changing the return type of the function! Better keep it like that and raise the exception in hoover instead.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 5]
```diff
@@ -301,7 +301,7 @@ def confirm_papers_to_author(pid, sigs_str, user_level=0):  # confirm_papers_to_
                 'and bibrec=%s',
                 (pid, '2', user_level, table, ref, rec))
     update_canonical_names_of_authors(pids_to_update)
-
+    #UPDATE EXTERNAL IDS OF AUTHOR!!
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:304]: ?

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 6]
```diff
@@ -129,7 +129,7 @@ def move_signature(sig, pid, force_claimed=False, set_unclaimed=False):
     if not force_claimed:
         query += " and (flag <> 2 and flag <> -2)"
 
-    run_sql(query, (pid, sig[0], sig[1], sig[2]))
+    run_sql(query, (pid, int(sig[0]), sig[1], sig[2]))
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:132]: Why was it changed? Have you checked the behaviour?

**Comment by kaplun** [comment]
Superseeds #128. @glouppe, @MSusik, @petros-ioannidis can you help reviewing API calls?

File: modules/bibauthorid/lib/bibauthorid_webapi.py [hunk 7]
```diff
@@ -3335,22 +3336,24 @@ def get_stored_incomplete_autoclaim_tickets(req):
     return temp_storage
 
 
-def add_cname_to_hepname_record(cname, recid, uid=None):
+def add_cname_to_hepname_record(cname_dict, uid=None):
     """
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_webapi.py:3340]: @petros-ioannidis change of API. Is it proofchecked?

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 1 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:793]: @petros-ioannidis  You completely changed the behaviour of the function -> previously it was only returning bibrecs, now it is returning everything.

You removed comments, and decided that you won't put any explanation. It seems that a person who developed this function before (maybe you) thought that it is not self-explanatory enough. Your code is not more readable than the previous one, so why didn't you write new comments?

There was a mode of this function, where you could ask for everything which was not rejected by authors. You removed it, I don't see a reason to do that.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 8]
```diff
@@ -2290,8 +2300,14 @@ def _get_inspire_id_of_author(pid):  # get_inspire_ids_by_pids
     @return: Inspire external identifier
     @rtype: tuple ((str),)
     '''
-    return _select_from_aidpersoniddata_where(select=['data'], pid=pid, tag='extid:INSPIREID')
-
+    result =  _select_from_aidpersoniddata_where(select=['data'], pid=pid, tag='extid:INSPIREID')
+    ##WRONG exception here
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2304]: then please fix it.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 3 above)
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2293]: Also, the type of the result is changed.

File: modules/bibauthorid/lib/bibauthorid_webapi.py [hunk 9]
```diff
@@ -3335,22 +3336,24 @@ def get_stored_incomplete_autoclaim_tickets(req):
     return temp_storage
 
 
-def add_cname_to_hepname_record(cname, recid, uid=None):
+def add_cname_to_hepname_record(cname_dict, uid=None):
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_webapi.py:3339]: Another change of the API. Please check the code for usages of this function.

**Comment by MSusik** [comment]
@petros-ioannidis Don't get it very personally, but your code that was merged caused a lot of problems with `bibauthorid`, `webauthorprofile` modules and my work in `miscutils`. Here are my suggestons:
- **You should check whole code for occurences of calls of functions that you are changing.** One of good programming practices is to make the code backward-compatible.
- If you didn't check behaviour of every component after changes, you should at least write tests for it.
- If you change crucial code, comment it.
- Write more tests (100% coverage might be too strict sometimes, but it is a good target to chase).
- Split the code into smaller, working commits. Here you could have prepared one commit with the changes for the API, and additional few Hoover commits.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 10]
```diff
@@ -2305,7 +2321,6 @@ def get_orcid_id_of_author(pid):  # get_orcids_by_pids
     '''
     return _select_from_aidpersoniddata_where(select=['data'], pid=pid, tag='extid:ORCID')
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2322]: This function was changed by you, not sure why the changes are not involved in this PR (maybe we changed it with Sam while fixing).

The type of the result was changed.

Please check `hoover` and the rest of the code for calls.

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py [hunk 11]
```diff
@@ -4105,8 +4121,23 @@ def bib_dictionarize(bibtable):
     MARC_100_700_CACHE['b100'] = bibd_10x
     MARC_100_700_CACHE['b700'] = bibd_70x
 
+    if (create_inverted_dicts or 'inverted_b100' in MARC_100_700_CACHE):
+        #MARC_100_700_CACHE['reversed_b100'] = dict((v,k) for k,v in MARC_100_700_CACHE['b100'].items())
+        MARC_100_700_CACHE['inverted_b100'] = dict(izip(MARC_100_700_CACHE['b100'].itervalues(), MARC_100_700_CACHE['b100'].iterkeys()))
+        #MARC_100_700_CACHE['reversed_b700'] = dict((v,k) for k,v in MARC_100_700_CACHE['b700'].items())
+        MARC_100_700_CACHE['inverted_b700'] = dict(izip(MARC_100_700_CACHE['b700'].itervalues(), MARC_100_700_CACHE['b700'].iterkeys()))
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:4128]: Please remove comments. It would be also nice to comment what is this section doing.

File: modules/bibauthorid/lib/bibauthorid_frontinterface.py [hunk 12]
```diff
@@ -183,7 +183,7 @@ def get_processed_external_recids(pid):
 
 
 def get_all_personids_recs(pid, claimed_only=False):
-    return dbinter.get_papers_of_author(pid, claimed_only)
+    return [x[3] for x in dbinter.get_papers_of_author(pid, not claimed_only)]
```
Comment (MSusik) [inline modules/bibauthorid/lib/bibauthorid_frontinterface.py:186]: I think it's very wrong.

File: modules/bibauthorid/lib/bibauthorid_daemon.py [hunk 13]
```diff
@@ -190,6 +231,22 @@ def _task_run_core():
...
+        query_file_size = bibtask.task_get_option("packet_size")
+        dry_hepnames = bibtask.task_get_option("dry_hepnames_run")
+        tickets = bibtask.task_get_option("open_tickets")
+        queue = bibtask.task_get_option("queue")
+        bibtask.task_update_progress('Initializing hoover')
+        if query_file_size:
+            run_hoover(check_db_consistency=check_db, dry_run=dry_run_b, packet_size=query_file_size, dry_hepnames_run=dry_hepnames, open_tickets=tickets)
+        if queue:
+            run_hoover(check_db_consistency=check_db, dry_run=dry_run_b, dry_hepnames_run=dry_hepnames, open_tickets=tickets, queue=queue)
+        else:
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_daemon.py:246]: Note, with this set of `if..else` if one sets `query_file_size` he will execute hoover over the default `Test` queue and then a second time!

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 8 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2304]: :tada:

File: modules/bibauthorid/lib/bibauthorid_dbinterface.py (hunk 10 above)
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_dbinterface.py:2322]: Yes, this was reverted in my latest OPS commit on top of this PR

File: modules/bibauthorid/lib/bibauthorid_hoover.py [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+        debug = ''
+    if rt_ticket_report:
+        if ticket_hash not in ticket_hashes.iterkeys():
+            ticket_id = BIBCATALOG_SYSTEM.ticket_submit(uid=None,
+                                                        subject=subject,
+                                                        recordid=e.recid,
+                                                        text=body +
+                                                        '\n Debugging\
+                                                        information: \n' +
+                                                        debug,
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_hoover.py:91]: Better to add the prefix `'\n Debugging information: \n'` the debug variable, so that if `debug_log` is `False`, there is no useless message.

File: modules/bibauthorid/lib/bibauthorid_hoover.py [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+    if not author_canonical_name:
+        return None
+
+    recid = perform_request_search(p="035:" + inspireid, cc="HepNames")
+    if recid:
+        if len(recid) > 1:
+            raise MultipleHepnamesRecordsWithSameIdException(
+                "More than one hepnames record found with the same inspire id",
+                recid,
+                'INSPIREID')
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_hoover.py:289]: Which INSPIREID? Better adding it to the exception.

File: modules/bibauthorid/lib/bibauthorid_hoover.py [hunk 16]
```diff
@@ -0,0 +1,746 @@
...
+def get_records_with_tag(tag):
+    """return all the records with a specific tag
+
+    Arguments:
+    tag -- the tag to search for
+    """
+    assert tag in ['100__i', '100__j', '700__i', '700__j']
+    if tag.startswith("100"):
+        return run_sql(
+            "select id_bibrec from bibrec_bib10x where id_bibxxx in (select id from bib10x where tag=%s)",
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_hoover.py:378]: Be careful: use the `search_engine.perform_request_search` API. It is as fast, and does not return e.g. deleted records or records not in HEP.

File: modules/bibauthorid/lib/bibauthorid_hoover.py [hunk 17]
```diff
@@ -0,0 +1,746 @@
...
+            (tag,
+             ))
+    if tag.startswith("700"):
+        return run_sql(
+            "select id_bibrec from bibrec_bib70x where id_bibxxx in (select id from bib70x where tag=%s)",
+            (tag,
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_hoover.py:384]: ditto

File: modules/bibauthorid/lib/bibauthorid_hoover.py [hunk 18]
```diff
@@ -0,0 +1,746 @@
...
+                'id_mapping': defaultdict(set)
+            }
+        },
+
+        "ORCID": {
+            'reliable': [  # get_orcid_id_of_author,
+                # get_inspireID_from_hepnames,
+                # lambda pid: get_inspireID_from_claimed_papers(pid,
+                # intersection_set=records_with_id)]
+            ],
```
Comment (kaplun) [inline modules/bibauthorid/lib/bibauthorid_hoover.py:561]: Why orcid is commented out :cry: . We should totally enable it. What is missing?
